test/bmv2_parsing_check.bzl

```diff
@@ -0,0 +1,67 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+def _subset_diff_impl(ctx):
+    # diff expected with tmp file
+    ctx.actions.write(
+        output = ctx.outputs.executable,
+        content = "{binary} < {input} | python {py_file} {input}".format(
+            py_file = ctx.file._py_file.short_path,
+            binary = ctx.file._binary.short_path,
+            input = ctx.file.input.short_path
+        )
+    )
+
+    runfiles = [ctx.file._py_file, ctx.file._binary, ctx.file.input]
+    return DefaultInfo(
+        runfiles = ctx.runfiles(files = runfiles),
+    )
+
+subset_diff_test = rule(
```

Comment:
It seems like this rule does two things: run the binary `main.cc`, and also compute a subset diff of the generated output with another json file. Can we refactor this so that it only does the second thing? I.e., it should take two json files, say `actual` and `expected` and compute the diff.
This will make your rule less complex and more reusable.

Simply use `genrule` to run the binary, no need to write a custom skylark rule for that.

Reply:
I sort of did this before reading this comment, the rules are now all split into coherent and self-containing pieces.

I have not used genrule, and instead used my own custom rule. I would like to keep it that way for two reasons:
1. The rule that I created autogenerates the output file names based on the input names, which is nice, because the user of the rule does not have to worry about paths and names etc

2. The rule outputs 2 files, and returns them using two different output groups (one json and one protobuf), which are easy to use as dependencies in other rules that require only a single file in a generic way. I do not know how this could be done with genrule, which gives all the output files together using the default provider.

For example, let us say I have defined some rule A responsible for running the binary and producing the json and pb.txt files, and now I want to feed only the json output file to some independent generic rule B defined separately.

if I just define B as
<something>(
   ...
   deps = [A]
)

and A is a plain gen rule, then B will receive both the json and pb.txt files via the default provider, which can cause errors if B is expecting a single file, or may do the wrong thing (e.g. apply B to both files instead of one etc).

The way I have it, the custom rule for A defines two output groups, then I can depend only on one of the two output groups in B's definition. This works even if B expects a single file, and does not know anything about the internals of A. It also does not restrict what providers rules are allowed to return if they are to be used in B.